### ISSUE TITLE
Fortran Bindings

### ISSUE THREAD

**Issue by bgoglin** [open-mpi/hwloc#190]
There is some interests in having Fortran bindings for basic features (such as binding threads or processes to cores or sockets or numa nodes).

The main reason is that there's no standard way to specify the binding in mpirun/mpiexec so people would rather spend time binding inside their application than learning multiple MPI implementation command-line options :)

**Comment by jsquyres** [comment]
FWIW, I'd argue that that's a terrible reason to add Fortran bindings to hwloc:

1. Rationale of "I'm too lazy to read documentation" is just weak sauce.
1. The MPI implementations already spent a lot of time and effort adding their binding functionality; throwing away all that effort is... odd.
1. That being said, if there's merit in trying to standardize some/all of the affinity mpirun options, then that might be a discussion worth having -- either between Open MPI / MPICH and/or at the Forum level.
1. Users will get the binding functionality wrong if they code it themselves -- it's nowhere near as simple as they think it is.
1. More specifically: it's quite naieve to think that writing code for effecting affinity is simpler than learning a few CLI options.
1. There are potential problems (and portability issues!) with linking hwloc in the user app if it's already in the MPI implementation (especially if the versions do not match).  I'm pretty sure that Open MPI will be fine, but I don't know if other implementations took as much care to hide hwloc properly.
1. There's also potential problems with overriding affinity/binding if the MPI has already done so (i.e., the MPI implementation's assumptions about the location of processes may now become incorrect/invalid, resulting in bad performance -- particularly with respect to shared memory latency).  The argument that you can have MPI not do the binding and therefore it shouldn't make assumptions about the process locality is moot because to do that, you have to read the MPI implementation's documentation to figure out which CLI options to use to disable affinity... at which point, why not read a few more paragraphs and figure out how to have the MPI implementation set the affinity that you want?

In short, I think "add Fortran bindings to hwloc so that I can do the binding myself" is analogous to "take that catalytic convertor out of my car so that I can convert the toxic gasses myself."  Meaning: it's needless and it's harder to do than you think.

However, if you do add Fortran bindings to hwloc, I'd (strongly) encourage two things:

1. Put them all in a "hwloc" Fortran module
2. Use the BIND(C) functionality so that Fortran is just calling the existing C bindings

This would mean that the "bindings" would essentially be declarations of the already-existing C functions in a Fortran module format.

**Comment by bgoglin** [comment]
closing as wontfix
